mono/metadata/icall.c

```diff
@@ -248,29 +248,32 @@ array_set_value_impl (MonoArrayHandle arr, MonoObjectHandle value, guint32 pos,
 	guint64 u64 = 0;
 	gint64 i64 = 0;
 	gdouble r64 = 0;
+	gboolean castOk = FALSE;
 
-	uint32_t arr_gchandle = 0;
-	uint32_t value_gchandle = 0;
+	MONO_ENTER_NO_SAFEPOINTS;
+
+	MonoArray *arr = (MonoArray *) MONO_HANDLE_RAW (arr_handle);
+	MonoObject *value = (MonoObject *) MONO_HANDLE_RAW (value_handle);
 
 	error_init (error);
 
-	if (!MONO_HANDLE_IS_NULL (value))
-		vc = mono_handle_class (value);
+	if (!MONO_HANDLE_IS_NULL (value_handle))
+		vc = mono_handle_class (value_handle);
 	else
 		vc = NULL;
 
-	ac = mono_handle_class (arr);
+	ac = mono_handle_class (arr_handle);
 	ec = m_class_get_element_class (ac);
 
 	esize = mono_array_element_size (ac);
-	ea = (gpointer*)mono_array_handle_pin_with_size (arr, esize, pos, &arr_gchandle);
+	ea = (gpointer*) mono_array_addr_with_size_internal (arr, esize, pos);
```

Comment:
maybe we can move this call closer to each place where `ea` is used and put smaller safepoint regions around each of those occurrences.